File: backend/internal/models/sync_backup.go
package models

import (
	"encoding/json"
	"time"
	"gorm.io/gorm"
)

// SyncJob 数据同步任务模型
type SyncJob struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name        string `gorm:"not null" json:"name"`
	Description string `json:"description"`
	
	// 同步源和目标
	SourcePath  string `gorm:"not null" json:"sourcePath"`
	DestPath    string `gorm:"not null" json:"destPath"`
	
	// 同步类型: local, remote (rclone), rsync
	Type        string `json:"type"` 
	
	// 计划配置
	Enabled     bool   `json:"enabled"`
	Schedule    string `json:"schedule"` // cron 表达式
	
	// 策略
	DeleteExtra bool   `json:"deleteExtra"` // --delete
	Checksum    bool   `json:"checksum"`    // --checksum
	
	// 状态
	Status      string     `json:"status"` // idle, running, failed, completed
	LastRun     *time.Time `json:"lastRun"`
	LastError   string     `json:"lastError"`
}

// BackupRepo Restic 备份仓库模型
type BackupRepo struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name string `gorm:"not null" json:"name"`
	// 仓库类型: local, s3, sftp, rest, b2, azure, gs, rclone
	Type string `gorm:"not null" json:"type"`
	// Restic 仓库规格（不含 scheme 部分），如 "/data/restic/repo1" 或 "s3.amazonaws.com/mybucket/path"
	URL string `json:"url"`
	// 完整的 restic -r 参数（由 Type+URL+其他参数拼成，运行时构造）
	// 留空则由后端自动拼接

	// 加密仓库密码（必填，restic init / 每次操作都需要）
	Password string `json:"password,omitempty"`
	// 凭据 / 连接参数（JSON 字符串）：根据后端不同填 access key/secret、SFTP 私钥、B2 keyID 等
	// 这些键会作为环境变量传给 restic（例如 AWS_ACCESS_KEY_ID、RESTIC_REPOSITORY 等）
	EnvJSON string `gorm:"type:text" json:"-"`

	// 起源：
	//   "" / "manual" — 用户在 UI 中手动创建
	//   "external:<container>" — 从外部 Docker 容器自动导入（绑定到该容器）
	// 当 Origin 是 external 时，立即备份 = docker start 对应容器，而不是本地 restic。
	Origin string `gorm:"default:'manual'" json:"origin"`

	// 状态: active / uninitialized / error
	Status      string     `gorm:"default:'uninitialized'" json:"status"`
	LastError   string     `gorm:"type:text" json:"lastError"`
	LastBackup  *time.Time `json:"lastBackup"`
	SnapshotCnt int        `json:"snapshotCount"`
	RepoSize    int64      `json:"repoSize"`
}

// MaskedEnv 返回脱敏后的环境变量（供前端展示）
func (r *BackupRepo) MaskedEnv() map[string]string {
	env := r.ParseEnv()
	out := make(map[string]string, len(env))
	for k, v := range env {
		if len(v) == 0 {
			continue
		}
		out[k] = maskValue(v)
	}
	return out
}

func maskValue(v string) string {
	if len(v) <= 4 {
		return "****"
	}
	return v[:2] + "****" + v[len(v)-2:]
}

// ParseEnv 解析 EnvJSON 为 map
func (r *BackupRepo) ParseEnv() map[string]string {
	out := map[string]string{}
	if r.EnvJSON == "" {
		return out
	}
	_ = json.Unmarshal([]byte(r.EnvJSON), &out)
	return out
}

// BackupTask Restic 备份任务
type BackupTask struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name string `gorm:"not null" json:"name"`
	// 关联仓库
	RepoID uint       `gorm:"not null" json:"repoId"`
	Repo   BackupRepo `gorm:"foreignKey:RepoID" json:"repo"`

	// 备份源路径（在容器内可访问的路径，例如 /host/data 或 /data/restic-data）
	SourcePath string `gorm:"not null" json:"sourcePath"`
	// 排除规则，每行一条（restic --exclude 语法）
	Excludes string `gorm:"type:text" json:"excludes"`
	// 标签（逗号分隔），用作 restic --tag
	Tags string `json:"tags"`
	// 保留策略 JSON，例如 {"keep_daily":7,"keep_weekly":4,"keep_monthly":12}
	// 留空表示不自动 forget
	RetentionJSON string `gorm:"type:text" json:"retention"`
	// 是否在备份完成后自动 prune
	AutoPrune bool `gorm:"default:false" json:"autoPrune"`

	// 触发模式：
	//   "" / "local"  — 在本容器内直接执行 restic backup
	//   "external"    — 通过 docker start 触发关联的外部容器（要求 Repo.Origin 以 external: 开头）
	TriggerMode string `gorm:"default:'local'" json:"triggerMode"`
	// 当 TriggerMode=external 时绑定的外部容器名；留空则用 Repo.Origin 的容器名
	ExternalContainer string `json:"externalContainer"`

	// 调度
	Enabled  bool   `gorm:"default:false" json:"enabled"`
	Schedule string `json:"schedule"` // cron 表达式（仅展示，目前由用户手动触发）

	// 状态字段
	Status         string     `gorm:"default:'idle'" json:"status"` // idle / running / completed / failed
	LastRun        *time.Time `json:"lastRun"`
	LastError      string     `gorm:"type:text" json:"lastError"`
	LastSnapshotID string     `json:"lastSnapshotId"`
	LastDuration   float64    `json:"lastDuration"` // 秒
}

// ParseRetention 解析保留策略
func (t *BackupTask) ParseRetention() map[string]int {
	out := map[string]int{}
	if t.RetentionJSON == "" {
		return out
	}
	// 用 encoding/json 解析 map[string]json.Number，再转 int
	var raw map[string]float64
	if err := json.Unmarshal([]byte(t.RetentionJSON), &raw); err == nil {
		for k, v := range raw {
			out[k] = int(v)
		}
	}
	return out
}

// BackupSyncJob 仓库到仓库的快照同步任务（restic copy）
type BackupSyncJob struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`

	Name string `gorm:"not null" json:"name"`
	// 源仓库
	SourceRepoID uint       `gorm:"not null" json:"sourceRepoId"`
	SourceRepo   BackupRepo `gorm:"foreignKey:SourceRepoID" json:"sourceRepo"`
	// 目标仓库
	TargetRepoID uint       `gorm:"not null" json:"targetRepoId"`
	TargetRepo   BackupRepo `gorm:"foreignKey:TargetRepoID" json:"targetRepo"`

	Enabled bool `gorm:"default:false" json:"enabled"`

	Status    string     `gorm:"default:'idle'" json:"status"`
	LastRun   *time.Time `json:"lastRun"`
	LastError string     `gorm:"type:text" json:"lastError"`
}

File: frontend/src/api/sync_backup.ts
// Restic Backup & Sync API Client
import api from './client'

// ===== Types =====
export interface BackupRepo {
  id: number
  name: string
  type: 'local' | 's3' | 'sftp' | 'rest' | 'b2' | 'azure' | 'gs' | 'rclone'
  url: string
  password?: string
  env?: Record<string, string>
  status: 'active' | 'uninitialized' | 'error'
  lastError?: string
  lastBackup?: string
  snapshotCount?: number
  repoSize?: number
  createdAt?: string
}

export interface ResticSnapshot {
  id: string
  short_id: string
  time: string
  paths: string[]
  hostname: string
  username?: string
  tags?: string[]
  summary?: {
    files_new: number
    files_changed: number
    files_unmodified: number
    total_files_processed: number
    total_bytes_processed: number
    data_added: number
  }
}

export interface BackupTask {
  id: number
  name: string
  repoId: number
  repo?: BackupRepo
  sourcePath: string
  excludes?: string
  tags?: string
  retention?: Record<string, number>
  autoPrune: boolean
  enabled: boolean
  schedule?: string
  status: 'idle' | 'running' | 'completed' | 'failed'
  lastRun?: string
  lastError?: string
  lastSnapshotId?: string
  lastDuration?: number
  createdAt?: string
}

export interface BackupSyncJob {
  id: number
  name: string
  sourceRepoId: number
  sourceRepo?: BackupRepo
  targetRepoId: number
  targetRepo?: BackupRepo
  enabled: boolean
  status: string
  lastRun?: string
  lastError?: string
}

export interface RepoStats {
  total_size: number
  total_blob_count: number
  snapshots_count: number
}

export interface LSNode {
  name: string
  type: 'file' | 'dir' | 'symlink'
  path: string
  size?: number
  permissions?: string
  mtime?: string
}

// ===== Ping / health =====
export interface PingInfo {
  ok: boolean
  error?: string
  hostname: string
  cacheDir?: string
  resticBin?: string
  version?: string
  runtime?: string
}

export interface BackupSettings {
  defaultHostname: string
  defaultExcludes: string
  defaultTags: string
  autoCheck: boolean
  confirmPurge: boolean
}

export interface WebDAVProfile {
  remoteName: string
  url: string
  vendor: string
  username: string
  password?: string
}

export const resticApi = {
  ping: () => api.get<PingInfo>('/api/storage/backup/ping'),

  // ===== Settings =====
  getSettings: () => api.get<BackupSettings>('/api/storage/backup/settings'),
  updateSettings: (data: BackupSettings) => api.put<BackupSettings>('/api/storage/backup/settings', data),

  // ===== WebDAV profile =====
  getWebDAV: () => api.get<WebDAVProfile>('/api/storage/backup/webdav'),
  updateWebDAV: (data: WebDAVProfile) => api.put<WebDAVProfile>('/api/storage/backup/webdav', data),
  testWebDAV: (data: WebDAVProfile) =>
    api.post<{ ok: boolean; error?: string; output?: string }>('/api/storage/backup/webdav/test', data),

  // ===== Repositories =====
  listRepos: () => api.get<{ repos: BackupRepo[]; total: number }>('/api/storage/backup/repos'),
  createRepo: (data: {
    name: string
    type: BackupRepo['type']
    url: string
    password: string
    env?: Record<string, string>
    init?: boolean
  }) => api.post<BackupRepo>('/api/storage/backup/repos', data),
  updateRepo: (id: number, data: Partial<{
    name: string
    type: BackupRepo['type']
    url: string
    password: string
    env?: Record<string, string>
  }>) => api.put<BackupRepo>(`/api/storage/backup/repos/${id}`, data),
  deleteRepo: (id: number, purge = false) =>
    api.delete(`/api/storage/backup/repos/${id}${purge ? '?purge=true' : ''}`),
  checkRepo: (id: number, full = false) =>
    api.post<{ output: string; status: string }>(`/api/storage/backup/repos/${id}/check${full ? '?full=true' : ''}`),
  refreshRepo: (id: number) =>
    api.post<{ status: string; stats?: RepoStats; repo?: BackupRepo }>(`/api/storage/backup/repos/${id}/refresh`),
  testRepo: (id: number) =>
    api.post<{ ok: boolean; error?: string; snapshotCount?: number; latestSnapshot?: string; status?: string }>(`/api/storage/backup/repos/${id}/test`),
  initRepo: (id: number) => api.post<BackupRepo>(`/api/storage/backup/repos/${id}/init`),
  unlockRepo: (id: number, force = false) =>
    api.post<{ output: string; status: string }>(`/api/storage/backup/repos/${id}/unlock${force ? '?force=true' : ''}`),

  // ===== Snapshots =====
  listSnapshots: (repoId: number) =>
    api.get<{ snapshots: ResticSnapshot[]; total: number }>(`/api/storage/backup/repos/${repoId}/snapshots`),
  snapshotDetail: (repoId: number, sid: string) =>
    api.get<{ snapshot: ResticSnapshot; fileCount: number; dirCount: number; totalSize: number }>(`/api/storage/backup/repos/${repoId}/snapshots/${sid}`),
  listSnapshotFiles: (repoId: number, sid: string) =>
    api.get<{ files: LSNode[]; total: number }>(`/api/storage/backup/repos/${repoId}/snapshots/${sid}/ls`),
  deleteSnapshot: (repoId: number, sid: string, prune = false) =>
    api.delete<{ output: string }>(`/api/storage/backup/repos/${repoId}/snapshots/${sid}${prune ? '?prune=true' : ''}`),
  diffSnapshots: (repoId: number, a: string, b: string) =>
    api.get<{ output: string }>(`/api/storage/backup/repos/${repoId}/diff?a=${a}&b=${b}`),
  findInSnapshots: (repoId: number, pattern: string) =>
    api.get<{ output: string; pattern: string }>(`/api/storage/backup/repos/${repoId}/find?q=${encodeURIComponent(pattern)}`),

  // ===== Restore =====
  restore: (repoId: number, data: {
    snapshotId?: string
    target: string
    include?: string[]
    exclude?: string[]
    host?: string
    paths?: string[]
  }) => api.post<{ message: string; logKey: string }>(`/api/storage/backup/repos/${repoId}/restore`, data),
  restoreLogs: (repoId: number) =>
    api.get<{ lines: string[] }>(`/api/storage/backup/repos/${repoId}/restore/logs`),

  // ===== Backup Tasks =====
  listTasks: () => api.get<{ tasks: BackupTask[]; total: number }>('/api/storage/backup/tasks'),
  createTask: (data: {
    name: string
    repoId: number
    sourcePath: string
    excludes?: string
    tags?: string
    retention?: Record<string, number>
    autoPrune?: boolean
    enabled?: boolean
    schedule?: string
  }) => api.post<BackupTask>('/api/storage/backup/tasks', data),
  updateTask: (id: number, data: any) => api.put<BackupTask>(`/api/storage/backup/tasks/${id}`, data),
  deleteTask: (id: number) => api.delete(`/api/storage/backup/tasks/${id}`),
  runTask: (id: number) => api.post<{ message: string; taskId: number }>(`/api/storage/backup/tasks/${id}/run`),
  taskStatus: (id: number) => api.get<{
    status: string; lastRun?: string; lastError?: string; lastSnapshotId?: string; lastDuration?: number
  }>(`/api/storage/backup/tasks/${id}/status`),
  taskLogs: (id: number) => api.get<{ lines: string[] }>(`/api/storage/backup/tasks/${id}/logs`),

  // ===== Repo-to-repo sync (restic copy) =====
  listSyncJobs: () => api.get<{ jobs: BackupSyncJob[]; total: number }>('/api/storage/backup/sync-jobs'),
  createSyncJob: (data: { name: string; sourceRepoId: number; targetRepoId: number; enabled?: boolean }) =>
    api.post<BackupSyncJob>('/api/storage/backup/sync-jobs', data),
  deleteSyncJob: (id: number) => api.delete(`/api/storage/backup/sync-jobs/${id}`),
  runSyncJob: (id: number) => api.post<{ message: string; jobId: number }>(`/api/storage/backup/sync-jobs/${id}/run`),
  syncJobLogs: (id: number) => api.get<{ lines: string[] }>(`/api/storage/backup/sync-jobs/${id}/logs`),
}

// ExternalContainer: 一个被发现的"外部 restic 容器"。
export interface ExternalContainer {
  id: string
  name: string
  image: string
  state: string
  status: string
  exitCode: number
  created: string
  startedAt?: string
  finishedAt?: string
  command: string
  composeProject?: string
  repo?: string
  sourcePath?: string
  retention?: string
  env: Record<string, string>
  mounts: { source: string; destination: string; mode: string }[]
}

// 保留旧版同步（rsync 文件级）API
export const syncApi = {
  getJobs: () => api.get('/api/storage/sync/jobs'),
  createJob: (data: any) => api.post('/api/storage/sync/jobs', data),
  runJob: (id: number) => api.post(`/api/storage/sync/jobs/${id}/run`),
}

// 旧版备份（tar/pg_dump）API；仅用于向下兼容的旧界面，新代码请使用 resticApi
export const backupApi = resticApi
